lib/pbench/server/auth/__init__.py

```diff
@@ -0,0 +1,313 @@
+from http import HTTPStatus
+from typing import Dict, List, Union
+from urllib.parse import urljoin
+
+import jwt
+import requests
+
+from pbench.server import JSON
+from pbench.server.auth.auth_provider_urls import (
+    URL_INTROSPECT,
+    URL_LOGOUT,
+    URL_REALM,
+    URL_TOKEN,
+    URL_USERINFO,
+    URL_WELL_KNOWN,
+)
+from pbench.server.auth.exceptions import KeycloakConnectionError
+
+
+class KeycloakOpenID:
+    """
+    Keycloak OpenID client.
+    :param server_url: Keycloak server url
+    :param client_id: client id
+    :param realm_name: realm name
+    :param client_secret_key: client secret key
+    :param verify: True if want check connection SSL
+    :param headers: dict of custom header to pass to each HTML request
+    """
+
+    def __init__(
+        self,
+        server_url: str,
+        realm_name: str,
+        client_id: str,
+        logger,
+        client_secret_key: str = None,
+        verify: bool = True,
+        headers: Dict = None,
```

Comment:
> any use of headers in the code will be flagged as a possible error since it might be None, so you really can't win...

I think my IDE is smart enough to do path-tracing, such that it only flags cases where it's actually possible that `headers` might be `None`.  (That is, if, at the top of the function, there is an `if not headers: headers = DEFAULT`, then I won't see any complaints; but, if the code is squirrelier, sometimes my IDE does produce a false positive.)